README.md
# agold/cljip

FIXME: my new application.

## Installation

Download from https://github.com/agold/cljip

## Usage

FIXME: explanation

Run the project directly, via `:exec-fn`:

    $ clojure -X:run-x
    Hello, Clojure!

Run the project, overriding the name to be greeted:

    $ clojure -X:run-x :name '"Someone"'
    Hello, Someone!

Run the project directly, via `:main-opts` (`-m agold.cljip`):

    $ clojure -M:run-m
    Hello, World!

Run the project, overriding the name to be greeted:

    $ clojure -M:run-m Via-Main
    Hello, Via-Main!

Run the project's tests (they'll fail until you edit them):

    $ clojure -T:build test

Run the project's CI pipeline and build an uberjar (this will fail until you edit the tests to pass):

    $ clojure -T:build ci

This will produce an updated `pom.xml` file with synchronized dependencies inside the `META-INF`
directory inside `target/classes` and the uberjar in `target`. You can update the version (and SCM tag)
information in generated `pom.xml` by updating `build.clj`.

If you don't want the `pom.xml` file in your project, you can remove it. The `ci` task will
still generate a minimal `pom.xml` as part of the `uber` task, unless you remove `version`
from `build.clj`.

Run that uberjar:

    $ java -jar target/cljip-0.1.0-SNAPSHOT.jar

If you remove `version` from `build.clj`, the uberjar will become `target/cljip-standalone.jar`.

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2021 Agold

_EPLv1.0 is just the default for projects generated by `deps-new`: you are not_
_required to open source this project, nor are you required to use EPLv1.0!_
_Feel free to remove or change the `LICENSE` file and remove or update this_
_section of the `README.md` file!_

Distributed under the Eclipse Public License version 1.0.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-12-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-12-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/agold/cljip/compare/0.1.1...HEAD
[0.1.1]: https://github.com/agold/cljip/compare/0.1.0...0.1.1


test/agold/cljip_test.clj
(ns agold.cljip-test
  (:require [clojure.test :refer [deftest testing is]]
            [agold.cljip :as a]
            #_[agold.ipgeo :as ipg]))




(deftest parse-log-test
  (testing "log parsing"
    (is (=
         (:ip (first (a/parse-log "/home/agold/Prog/cljip/testdata/default.log")))
         "39.104.69.228"))))


test/agold/ipgeo_test.clj
(ns agold.ipgeo-test
  (:require [clojure.test :refer [deftest testing is]]
            #_[agold.cljip :as a]
            [agold.ipgeo :as ipg]))

(deftest config-test
  (testing "config loading"
    (is (not (nil? (:API-KEY (ipg/get-config)))))))

(deftest dns-test
  (testing "reverse dns if host exists"
    (is (= (ipg/get-hostname "100.35.79.95")
           {:hostname "pool-100-35-79-95.nwrknj.fios.verizon.net"})))
  (testing "reverse dns if host does not exist"
    (is (= (ipg/get-hostname "190.35.79.95")
           {:hostname "190.35.79.95"}))))

test/agold/dateparser_test.clj
(ns agold.dateparser-test
  (:require [clojure.test :refer [deftest testing is]]
            [agold.dateparser :as dp])
  (:import (java.time ZonedDateTime)))

(deftest jtime-conv-test
  (testing "conversion of log date format to jtime zoned date"
    (is (= (type (dp/datestr->jtime "27/Feb/2021:00:58:22"))
           ZonedDateTime))))

src/agold/ip-pp.clj
(ns agold.ip-pp
  (:require [clojure.core.async :as a]
            [clojure.pprint :as pp]))

#_(defn pp-le-from-chan
  "pretty print log entry from channel"
  [pch]
  (a/go (pp/pprint (a/<! pch))))

(defn start-print-loop
  "create print channel for pretty printing log entries"
  []
  (let [prn-chan (a/chan)]
    (a/go-loop [item (a/<! prn-chan)]
      (when item
        (pp/pprint item)
        (recur (a/<! prn-chan))))
    prn-chan))

(comment
  (def smpl-le {:entry
                "47.241.66.187 - - [27/Feb/2021:13:30:39 +0000] \"GET / HTTP/1.1\""
                :ip "47.241.66.187"
                :date
                "date object here"
                :req "GET / HTTP/1.1"
                :site-data
                {:country_code2 "SG"
                 :city "Singapore"
                 :longitude "103.85098"
                 :zipcode "048616"
                 :country_name "Singapore"
                 :country_code3 "SGP"
                 :latitude "1.28434"
                 :state_prov ""
                 :district "Central Business District"}})
  (let [pch (start-print-loop)]
    (a/>!! pch smpl-le))
  )


src/agold/ip_pp.clj
(ns agold.ip-pp
  (:require [clojure.core.async :as a]
            [clojure.pprint :as pp]))

#_(defn pp-le-from-chan
  "pretty print log entry from channel"
  [pch]
  (a/go (pp/pprint (a/<! pch))))

(defn start-print-loop
  "create print channel for pretty printing log entries"
  []
  (let [prn-chan (a/chan 2048)]
    (a/go-loop [item (a/<! prn-chan)]
      (when item
        (pp/pprint item)
        (recur (a/<! prn-chan))))
    prn-chan))

(comment
  (def smpl-le {:entry
                "47.241.66.187 - - [27/Feb/2021:13:30:39 +0000] \"GET / HTTP/1.1\""
                :ip "47.241.66.187"
                :date
                "date object here"
                :req "GET / HTTP/1.1"
                :site-data
                {:country_code2 "SG"
                 :city "Singapore"
                 :longitude "103.85098"
                 :zipcode "048616"
                 :country_name "Singapore"
                 :country_code3 "SGP"
                 :latitude "1.28434"
                 :state_prov ""
                 :district "Central Business District"}})
  (let [pch (start-print-loop)]
    (a/>!! pch smpl-le))
  )


src/agold/ipgeo.clj
(ns agold.ipgeo
  (:require [clojure.core.async :as a]
            [config.core :as e])
  (:import [java.net InetAddress]))

;; using app.ipgeolocation.io for ip lookup
"
 # Get geolocation for an IPv4 IP Address = 8.8.8.8
$ curl 'https://api.ipgeolocation.io/ipgeo?apiKey=API_KEY&ip=8.8.8.8'

# Get geolocation for an IPv6 IP Address = 2001:4860:4860::1
$ curl 'https://api.ipgeolocation.io/ipgeo?apiKey=API_KEY&ip=2001:4860:4860::1'

# Get geolocation for a domain name = google.com
$ curl 'https://api.ipgeolocation.io/ipgeo?apiKey=API_KEY&ip=dns.google.com
 
 for fields selection:
 $ curl 'https://api.ipgeolocation.io/ipgeo?apiKey=API_KEY&ip=1.1.1.1&fields=city'
"
(defn get-config
  "load config from config.edn in classpath"
  []
  (e/load-env))


(defonce API-KEY (:API-KEY (get-config)))

(defn reverse-dns-lookup
  "homespun version of reverse dns"
  [ip]
  (.getCanonicalHostName (InetAddress/getByName ip)))

(defn get-hostname
  "get host from ip address, handle not found exception"
  [ip]
  (let [hostname (atom "none")]
    (try
      (reset! hostname (#(reverse-dns-lookup ip)))
      (catch Exception _ (#(reset! hostname "Host Not Found"))))
    {:hostname @hostname}))

(def get-hostname-cached (memoize get-hostname))

(defn async-get-hostname
  "get hostname asynchronously and put to result channel
   with timeout max-delay in ms (default 100)"
  ([ip resch]
   (async-get-hostname ip resch 100))
  ([ip resch max-delay]
   #_(println "agh called")
   (let [hnch (a/chan 1)
         toch (a/timeout max-delay)]
     (a/go
       (a/>! hnch (get-hostname-cached ip))
       (a/alt!
         toch (a/>! resch {:hostname "Timed out"})
         hnch ([hn] (a/>! resch hn)))))))

;; TODO: for testing, remove later
(defn try-async
  [ip delay]
  (let [ch (a/chan 50)]
    (async-get-hostname ip ch delay)
    ch))

(comment
  (reverse-dns-lookup "8.8.8.8")
  (reverse-dns-lookup "47.241.66.187")
  (get-hostname "47.241.66.187")
  (get-hostname-cached "47.241.66.187")
  (get-hostname-cached "100.35.79.95")
  (get-hostname-cached "8.8.8.8")
  ;; pool-....
  (get-hostname-cached "190.35.79.95")
  ;; host not found, will return numerical ip addr as string
  (get-hostname-cached "71.192.181.208")
  (def try (try-async "106.36.79.95" 20))

  (.count (.buf try))
(println (a/<!! try))
  )

src/agold/dateparser.clj
(ns agold.dateparser
  #_{:clj-kondo/ignore [:unused-import]}
  (:import (java.time ZonedDateTime format.DateTimeFormatter)))

#_{:clj-kondo/ignore [:unresolved-namespace]}
(def formatter (DateTimeFormatter/ofPattern "yyyyMMdd:HHmmss z"))

(def month-map
  {"Jan" "01" "Feb" "02" "Mar" "03" "Apr" "04" "May" "05" "Jun" "06"
   "Jul" "07" "Aug" "08" "Sep" "09" "Oct" "10" "Nov" "11" "Dec" "12"})

(get month-map "Jan")

(defn- datestr->yyyyMMdd
  "converts date string 27/Feb/2021:00:58:22 from log to yyyyMMdd:hhMMss string"
  [datestr]
  (let [;;re #"(\S+)/(\S+)/(\S+?):"
        re #"(\S+)/(\S+)/(\S+?):(\S+):(\S+):(\S+)"
        parsed (re-find re datestr)]
    (str (parsed 3) (get month-map (parsed 2)) (parsed 1) ":" (parsed 4) (parsed 5)
         (parsed 6) " GMT+00:00")))

(defn datestr->jtime
  "converts log date str 27/Feb/2021:00:58:22 to Java date object"
  [datestr]
  (ZonedDateTime/parse (datestr->yyyyMMdd datestr) formatter))

(comment
  (def datestr "27/Feb/2021:00:58:22")
  (re-find #"(\S+)/(\S+)/(\S+?):(\S+):(\S+):(\S+)" datestr)
  (datestr->yyyyMMdd "27/Feb/2021:00:58:22")
  (ZonedDateTime/parse "20210727:005822 GMT+00:00" formatter)
  (datestr->jtime "27/Feb/2021:00:58:22")
  )

src/agold/cljip.clj
(ns agold.cljip
  (:require [clojure.core.async :as a]
            [clojure.java.io :as io]
            [clojure.pprint :as pp]
            [clojure.string :as string]
            [cheshire.core :as ch]
            [org.httpkit.client :as http]
            [agold.dateparser :as dp]
            [agold.ipgeo :as ipg]
            [agold.ip-pp :as ipp])
  (:gen-class))

(defonce ipgeo "https://api.ipgeolocation.io/ipgeo")


(defn get-site-data
  "fetch reverse dns and geodata for ip addr"
  [ip]
  (let [url (str ipgeo "?apiKey=" ipg/API-KEY "&ip=" ip "&fields=geo")]
    #_:clj-kondo/ignore
    (let [resp @(http/get url)]
      (if (= (:status resp) 200)
        {:site-data (dissoc (ch/parse-string (:body resp) true) :ip)}
        {:site-data "N/A"}))))

(def get-site-data-cached (memoize get-site-data))

(defn log->vec-of-lines
  "log file to vector of lines"
  [fname]
  (let [lines []]
    (with-open [rdr (io/reader fname)]
      (into lines (line-seq rdr)))))

(defn parse-line
  "parse line of log"
  [line]
  (let [parse-re #"(\S+).+?[\[](\S+).+?\"(.+?)\""
        parsed (re-find parse-re line)]
    {:entry (parsed 0)
     :ip (parsed 1)
     :date (parsed 2)
     :req (parsed 3)}))

;; log-entry has shape {:entry x :ip y :date d :req r}

(defn fix-date
  "get date part of logentry"
  [log-entry]
  (let [date (:date log-entry)]
    #_(println date)
    (assoc log-entry :date (dp/datestr->jtime date))))

(defn parse-log
  "logfname to vec of logentries with jdatetimes"
  [logfname]
  (let [vec-logentries (mapv parse-line (log->vec-of-lines logfname))]
    (mapv fix-date vec-logentries)))

#_(defn add-hostname
    "adds hostname to log entry"
    [le ch]
    (merge le (ipg/async-get-hostname (:ip le) midch 20)))

(defn add-site-data
  "add site data to log entry"
  [le]
  (merge le (get-site-data-cached (:ip le))))

(defn add-site-data-async
  "add site data asynchronously"
  [le result]
  (a/go
    (a/>! result (add-site-data le))
     ;; TODO removing below line produces only first log entry in seq, why??
    (a/close! result)))

(defn add-hostname-async
  "add hostname asynchronously"
  [le result]
  (let [hn-chan (a/chan 20)]
    (a/go
     ;; get hostname with max 20 ms delay
      (ipg/async-get-hostname (:ip le) hn-chan 20)
      (a/>! result (merge le (a/<! hn-chan)))
      #_(a/>! result (merge le {:hostname "dummy"}))
     ;; TODO removing below line produces only first log entry in seq, why??
      (a/close! result))))

(defn add-data-async
  "add data to log entries on input channel inch, pass to outch"
  [inch midch outch]

  (a/pipeline-async 8 midch add-hostname-async inch)
  (a/pipeline-async 8 outch add-site-data-async midch)
  #_(a/pipeline-async 8 outch add-site-data-async inch))

(defn process-log
  "process log file"
  [logfname]
  (let [vec-of-les (parse-log logfname)
        outch (ipp/start-print-loop)
        midch (a/chan 2048)
        inch (a/chan 2048)]
    (add-data-async inch midch outch)
    #_(a/go-loop [item (a/<! outch)]
        (when item
          (println item)
          (recur (a/<! outch))))
    (a/go (doseq [le vec-of-les]
            #_(println "adding ...")
            (a/>! inch le)))
    :done))

#_:clj-kondo/ignore
(defn proclog
  "Callable entry point to the application."
  [data]
  (println "conf key is: " (:API-KEY (ipg/get-config)))
  (println "Starting, exit after 10 secs")
  (process-log "testdata/newer.log")
  (a/<!! (a/timeout 10000))
  (println "exiting")
  (println (str "Bye, " (or (:file data) "World") "!")))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println (str "Starting:" (or args "No args")))
  (process-log "testdata/newer.log")
  (println "exiting"))

#_:clj-kondo/ignore
(comment
  (greet {:name "Art"})
  (string/lower-case "FOO")
  (get-site-data "8.8.8.8")
  (get-site-data "71.192.181.208")
  ;; https://github.com/apribase/clj-dns/blob/master/src/clj_dns/core.clj
  ;; https://gist.github.com/mwchambers/1316080
  (.getCanonicalHostName (InetAddress/getByName "100.35.79.95"))
  (ipg/get-hostname-cached "8.8.8.8")
  (ipg/get-hostname-cached "47.241.66.187")
  ;; (jn/InetAddress.get-host-address "8.8.8.8")
  (re-find #"\".+" "Hello \"Dolly\"")
  (def logstr "180.95.238.249 - - [27/Feb/2021:01:04:43 +0000] \"GET http://www.soso.com/ HTTP/1.1\" 200 396 \"-\" \"Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.101 Safari/537.36\"")
  (re-find #"(\S+).+[\[](\S+).+?\"(.+?)\"" logstr)
  (System/getProperty "user.home")
  (log->vec-of-lines "/home/agold/Prog/cljip/testdata/default.log")
  (println (parse-line logstr))
  (time
   (map parse-line (log->vec-of-lines "/home/agold/Prog/cljip/testdata/default.log")))
  (time (parse-log "/home/agold/Prog/cljip/testdata/default.log"))
  (def default-log "/home/agold/Prog/cljip/testdata/default.log")
  (map add-hostname (parse-log default-log))
  (let [procd (map add-site-data (parse-log default-log))]
    (doseq [le procd]
      (pp/pprint le)))
  (process-log default-log)
  (process-log "testdata/newer.log")
  (def p-ch (a/chan))
  (a/go (println (a/<! p-ch)))
  (a/go (a/>! "Hello"))
  (a/go-loop [item (a/<! p-ch)]
    (when item
      (println item)
      (recur (a/<! p-ch))))
  (a/>!! p-ch "Hello Dolly")
  (println (slurp "config.edn"))
  ;; gets current working directory
  (.getCanonicalFile (clojure.java.io/file "."))
  (-main ""))





